packages/hardhat-core/src/internal/hardhat-network/provider/vm/dual.ts

```diff
@@ -89,42 +100,102 @@ export class DualModeAdapter implements VMAdapter {
   }
 
   public async getStateRoot(): Promise<Buffer> {
-    return this._ethereumJSAdapter.getStateRoot();
+    const ethereumJSRoot = await this._ethereumJSAdapter.getStateRoot();
+    const rethnetRoot = await this._rethnetAdapter.getStateRoot();
+
+    if (!ethereumJSRoot.equals(rethnetRoot)) {
+      console.trace(
+        `Different state root: ${ethereumJSRoot.toString(
+          "hex"
+        )} !== ${rethnetRoot.toString("hex")}`
+      );
+      throw new Error("Different state root");
+    }
+
+    return rethnetRoot;
   }
 
   public async getAccount(address: Address): Promise<Account> {
-    return this._ethereumJSAdapter.getAccount(address);
+    const ethereumJSAccount = await this._ethereumJSAdapter.getAccount(address);
+    const rethnetAccount = await this._rethnetAdapter.getAccount(address);
+
+    assertEqualAccounts(address, ethereumJSAccount, rethnetAccount);
+
+    return ethereumJSAccount;
   }
 
   public async getContractStorage(
     address: Address,
     key: Buffer
   ): Promise<Buffer> {
-    return this._ethereumJSAdapter.getContractStorage(address, key);
+    const ethereumJSStorageSlot =
+      await this._ethereumJSAdapter.getContractStorage(address, key);
+
+    const rethnetStorageSlot = await this._rethnetAdapter.getContractStorage(
+      address,
+      key
+    );
+
+    if (
+      !(
+        ethereumJSStorageSlot.length === 0 &&
+        rethnetStorageSlot.equals(Buffer.from([0x00]))
+      ) &&
+      !ethereumJSStorageSlot.equals(rethnetStorageSlot)
+    ) {
+      console.trace(
+        `Different storage slot: ${bufferToHex(
+          ethereumJSStorageSlot
+        )} !== ${bufferToHex(rethnetStorageSlot)}`
+      );
+      throw new Error("Different storage slot");
+    }
+
+    return rethnetStorageSlot;
   }
 
   public async getContractCode(address: Address): Promise<Buffer> {
-    return this._ethereumJSAdapter.getContractCode(address);
+    const ethereumJSCode = await this._ethereumJSAdapter.getContractCode(
+      address
+    );
+    const rethnetCode = await this._rethnetAdapter.getContractCode(address);
+
+    if (!ethereumJSCode.equals(rethnetCode)) {
+      console.trace(
+        `Different contract code: ${ethereumJSCode.toString(
+          "hex"
+        )} !== ${rethnetCode.toString("hex")}`
+      );
+      throw new Error("Different contract code");
+    }
+
+    return rethnetCode;
   }
 
   public async putAccount(address: Address, account: Account): Promise<void> {
-    return this._ethereumJSAdapter.putAccount(address, account);
+    await this._ethereumJSAdapter.putAccount(address, account);
+    return this._rethnetAdapter.putAccount(address, account);
```

Comment:
```suggestion
    await this._rethnetAdapter.putAccount(address, account);
```

Personal preference: I would await both calls instead of returning the second. Otherwise, this makes you stop to think "why is the second value returned", when it's not necessary.

Same for the other similar methods.

Reply:
Will update